### ISSUE TITLE
Benchmark results on Sana 1600M + open perf/accuracy issues

### ISSUE THREAD

**Issue by florianmattana** [ParagEkbote/model-kernels#11]
Posting benchmark results from PR #8 here for tracking, plus two open issues 
worth discussing.

## Setup

- GPU: RTX 5070 Ti Laptop (SM120, 12 GB)
- Model: Efficient-Large-Model/Sana_1600M_1024px_diffusers
- dtype: bf16
- Steps: 4, Runs: 5, Warmup: 2

## Results

| Config    | Mean    | P50     | P95     | Mem (GB) | vs SDPA |
|-----------|---------|---------|---------|----------|---------|
| default   | 7.508s  | 7.508s  | 7.510s  | 10.91    | 0.67×   |
| sdpa_all  | 5.047s  | 6.674s  | 6.678s  | 10.91    | 1.00×   |
| int8      | 9.750s  | 9.753s  | 9.815s  | 10.91    | 0.52×   |

Mean pixel diff (INT8 vs sdpa_all): 107.7 / 255.

## Side note on the benchmark setup

The Sana wrapper in benchmark_sana_int8.py was crashing with a segfault on the 
int8 config. Root cause was a schema contract violation: SanaInt8AttnProcessor 
was passing `self._no_ts = torch.Tensor()` (or a size-1 tensor) as 
timestep_scales, but validate_timestep_scales in torch_binding.cpp requires 
`ts.size(0) == batch_size`. The crash manifested as a silent segfault rather 
than a RuntimeError, probably because the TORCH_CHECK failure happens inside 
a custom op dispatch boundary.

Fix: pass `None` instead, which skips validation via `has_value() == false` 
and the kernel uses `ts=1.f` internally. One-line change in 
`SanaInt8AttnProcessor.__init__`.

## Open issue 1 — Performance

INT8 kernel is ~1.9× slower than SDPA on this workload. Part of that is that 
SDPA on Ampere+ dispatches to FlashAttention-2 which is hard to beat on 
seq_len=1024, but looking at the online softmax loop in attention_int8.cu:

```cpp
for (int qi = wid; qi < q_size; qi += WARPS) {
    ...
    for (int ki = 0; ki < k_size; ++ki) { ... }
    for (int d = 0; d < HEAD_DIM; ++d) {
        out_acc[qi * HEAD_DIM + d] *= rescale;
    }
    ...
}
```

The per-row work is partitioned by warp but not by lane. All 32 threads in a 
warp execute the same loop over `ki` and `d`, compute the same values, and 
write to the same `out_acc` addresses. Splitting the `d` loop across lanes 
and using warp shuffles for the per-row max and sum reductions would likely 
recover a significant chunk of the gap. Happy to open a separate PR for this 
if it's useful.

## Open issue 2 — Accuracy on head_dim=32

Mean pixel diff of 107.7/255 between INT8 and sdpa_all is not a degradation, 
the generated image is essentially different. Sana 1600M uses head_dim=32 for 
self-attention, which is technically in the supported list but may not have 
been tested against real activation distributions. Possible culprits:

- Dynamic per-tile quantization on very narrow head_dim may be hitting 
  outlier values that push the INT8 range past what Sana's activations 
  actually need
- The global K scale fix from PR #8 was validated on synthetic tensors at 
  N ∈ {64, 512, 2048}; real Sana activations may have a different distribution
- Possible interaction with Sana's linear-attention-style normalization that 
  I haven't dug into

I'd expect the kernel to be much cleaner on head_dim ∈ {64, 128}. Might be 
worth benchmarking on a model with larger head_dim (SD3, a smaller DiT) to 
confirm whether this is a head_dim=32 corner case or a deeper correctness 
issue.

**Comment by ParagEkbote** [comment]
Quite interesting set of points, firstly why do you think we are not observing reduction in VRAM usage while applying the kernel. For a larger head_dim, for this we would require more compute, right? Next, what script have you used to get the benchmark values, could you share it with me and the installation and usage steps for the kernel remain same as written in this [PR comment](https://github.com/ParagEkbote/model-kernels/pull/8#issuecomment-4080475211), right?

cc: @florianmattana

**Comment by florianmattana** [comment]
For the vram the thing is that the int8 representation only exists inside the kernel's shared memory during execution, it never lands in PyTorch's allocator.  Q, K, V stay bf16 in global memory the whole time, and the output is allocated with`torch::empty_like(Q)` so it inherits the bf16 dtype  too. From the allocator's point of view, nothing has changed compared  to SDPA, so `max_memory_allocated()` reports the same number. On top of that, the peak memory in this benchmark isn't set by  attention at all. It's set by Sana's VAE decoder at the end of the pipeline, which upscales the latent to full image resolution and dwarfs the attention working set.

For the head_dim qustion id say this is the opposite no ? The inner matmul loop runs `D / 16` Tensor Core instructions per sharedmemory tile. At head_dim=32 that's only 2 instructions per load, which barely warms up the Tensor Core pipeline. At head_dim=128 it's 8, much better arithmetic intensity, more compute per byte loaded.

The dynamic per-tile quantization is also more stable at larger head_dim. With more values per row setting the max, the scale is less sensitive to a single outlier, so the quantization error stays bounded. At head_dim=32 a single outlier can dominate the scale and 
collapse everything else. So head_dim=32 is roughly the worst case for this kernel on both  fronts at once: not enough work to amortize the quantization, and a scale that's too easy to skew

Edit ( forgot to answer for the script) : 
same one you posted, install and usage steps unchanged. The only modification is in SanaInt8AttnProcessor.__init__, replacing the placeholder timestep_scales tensor with None
